extensions/amp-ad-exit/0.1/amp-ad-exit.js

```diff
@@ -136,13 +151,23 @@ export class AmpAdExit extends AMP.BaseElement {
             const vendorResponse = replacements./*OK*/expandStringSync(
```

Comment:
1. `!(customVar = /** @type {!./config.VariableDef} */ (target.vars[customVarName]))` is just code golf. Please split into multiple if statements, with a variable assignment in between.
2. I don't understand why this is being done this way. You have `n` nested calls to expand, when you could have just added the `IFRAME_TRANSPORT_SIGNAL` as a custom replace function in the original?

Reply:
1. I can happily change it back. Please note that I had it that way originally and changed to this due to code review comment (https://github.com/ampproject/amphtml/pull/11462#pullrequestreview-65664585) - it would be good if we could ammend some spec to address this.

2. I fully agree. Honestly I don't like this IFRAME_TRANSPORT_SIGNAL(vendor, key) format at all, and would prefer to just specify JSON values directly: { ..., vendor: my-vendor, key: my-key, ... }.  As you note, that would avoid a ton of code. But, I lost that battle (see design doc comment here: https://docs.google.com/document/d/1gSRS1U17dwRM-uFIuk8xp7bYHJ1acsMOHPvN-NTYpZw/edit#)
To clarify, it seems that we kind of have to do a two-step replacement: _someVar -> "the value that vendor X sent for key Y" -> whatever that value actually is.
So, that being said...I don't know a better way to do this, with the caveat of using IFRAME_TRANSPORT_SIGNAL. Any advice?

Reply:
1. Done

Reply:
@lannka: Doesn't analytics have a recursive replace somewhere? Maybe we can reuse.

Reply:
analytics don't have recursive replacement function. We achieve recursive replacement by calling `urlReplacementsForDoc.expandUrlAsync` at different places.

Reply:
[This](https://github.com/ampproject/amphtml/blob/master/extensions/amp-analytics/0.1/variables.js#L191-L253) is what I was thinking of.

Reply:
Oh!`VariableExpansion` should work to replace `responseKey` value to the real response. We can generate the `ExpansionOptions` based on the returned value from 3p vendor iframe.
However I think it's unnecessary here, 3p vendor iframe will not return variables recursively. What we have here should be sufficient.

Reply:
> 3p vendor iframe will not return variables recursively.

It can, though, to depth 2 (the line this comment thread is on). The `IFRAME_TRANSPORT_SIGNAL` is the first depth, and it may or may not return variables that need to be replaced themselves.

Reply:
Once we replace `${_bar}` with `IFRAME_TRANSPORT_SIGNAL(example 3p vendor, collected-data)`. `IFRAME_TRANSPORT_SIGNAL` is supposed to be a substitutionFunction. variable expansion no longer works but a substitution function like what we have now should be implemented.

Reply:
I appreciate all the tips here! Created https://github.com/ampproject/amphtml/issues/11623 to improve this.